storage/filesystem/object.go

```diff
@@ -22,43 +22,49 @@ type ObjectStorage struct {
 
 func newObjectStorage(dir *dotgit.DotGit) (ObjectStorage, error) {
 	s := ObjectStorage{
-		dir:   dir,
-		index: make(map[plumbing.Hash]idx, 0),
+		dir: dir,
 	}
 
-	return s, s.loadIdxFiles()
+	return s, nil
 }
 
-func (s *ObjectStorage) loadIdxFiles() error {
+func (s *ObjectStorage) requireIndex() error {
+	if s.index != nil {
+		return nil
+	}
+
+	s.index = make(map[plumbing.Hash]idx, 0)
 	packs, err := s.dir.ObjectPacks()
 	if err != nil {
 		return err
 	}
 
 	for _, h := range packs {
-		if err := s.loadIdxFile(h); err != nil {
+		idxfile, err := s.dir.ObjectPackIdx(h)
```

Comment:
because with this you can hide some complexity in this function, also this will help to simplify the code, putting the close in a defer, and inline err check.